sys/net/gnrc/link_layer/lwmac/lwmac_internal.c

```diff
@@ -0,0 +1,298 @@
+{
+    /* Counter did overflow since last wakeup */
+    if (rtt_get_counter() < last) {
+        /* TODO: Not sure if this was tested :) */
+        uint32_t tmp = -last;
+        tmp /= interval;
+        tmp++;
+        last += tmp * interval;
+    }
+
+    /* Add margin to next wakeup so that it will be at least 2ms in the future */
+    while (last < (rtt_get_counter() + LWMAC_RTT_EVENT_MARGIN_TICKS)) {
+        last += interval;
+    }
+
+    return last;
+}
+
+void lwmac_print_hdr(lwmac_hdr_t *hdr)
+{
+    assert(hdr != NULL);
+
+    printf("LwMAC header:\n  Type: ");
+    switch (hdr->type) {
+        case FRAMETYPE_WR: {
+            puts("Wakeup request (WR)");
+            break;
+        }
+        case FRAMETYPE_WA: {
+            puts("Wakeup acknowledge (WA)");
+            printf("  Src addr:");
+            lwmac_frame_wa_t *wa = (lwmac_frame_wa_t *) hdr;
+            for (int i = 0; i < wa->dst_addr.len; i++) {
+                printf("0x%02x", wa->dst_addr.addr[i]);
+                if (i < (wa->dst_addr.len - 1)) {
+                    printf(":");
+                }
+            }
+            break;
+        }
+        case FRAMETYPE_DATA: {
+            puts("User data");
+            break;
+        }
+        case FRAMETYPE_BROADCAST: {
+            puts("Broadcast user data");
+            printf("  Sequence number: %d\n", ((lwmac_frame_broadcast_t *)hdr)->seq_nr);
+            break;
+        }
+        default: {
+            puts("Unkown type");
+            printf("  Raw:  0x%02x\n", hdr->type);
+        }
+    }
+}
+
+int _dispatch_defer(gnrc_pktsnip_t *buffer[], gnrc_pktsnip_t *pkt)
+{
+    assert(buffer != NULL);
+    assert(pkt != NULL);
+
+    /* We care about speed here, so assume packet structure */
+    assert(pkt->next->type == GNRC_NETTYPE_LWMAC);
+    assert(pkt->next->next->type == GNRC_NETTYPE_NETIF);
+
+    lwmac_frame_broadcast_t *bcast = NULL;
+    if (((lwmac_hdr_t *)pkt->next->data)->type == FRAMETYPE_BROADCAST) {
+        bcast = pkt->next->data;
+    }
+
+    for (unsigned i = 0; i < GNRC_MAC_DISPATCH_BUFFER_SIZE; i++) {
+        /* Buffer will be filled bottom-up and emptied completely so no holes */
+        if (buffer[i] == NULL) {
+            buffer[i] = pkt;
+            return 0;
+        }
+        else {
+            /* Filter same broadcasts, compare sequence number */
+            if (bcast &&
+                (((lwmac_hdr_t *)buffer[i]->next->data)->type == FRAMETYPE_BROADCAST) &&
+                (bcast->seq_nr == ((lwmac_frame_broadcast_t *)buffer[i]->next->data)->seq_nr)) {
+                gnrc_netif_hdr_t *hdr_queued, *hdr_new;
+                hdr_new = pkt->next->next->data;
+                hdr_queued = buffer[i]->next->next->data;
+
+                /* Sequence numbers match, compare source addresses */
+                if (hdr_new->src_l2addr_len == hdr_queued->src_l2addr_len) {
+                    if (memcmp(gnrc_netif_hdr_get_src_addr(hdr_new),
```

Comment:
```C
if ((hdr_new->src_l2addr_len == hdr_queued->src_l2addr_len) &&
    (memcmp(gnrc_netif_hdr_get_src_addr(hdr_new),
            gnrc_netif_hdr_get_src_addr(hdr_queued),
            hdr_new->src_l2addr_len) == 0)) {
...
```

More saving of indentation levels and saving a jump instruction in compiled code ;-)